### PR TITLE
Add support for Pytorch training

### PR DESCRIPTION
**Pull request by ealt**

### CODE REVIEW

File: simplexity/evaluation/evaluate_pytorch_model.py
Comment (hrbigelow) [inline simplexity/evaluation/evaluate_pytorch_model.py:1]: Overall I think a user of simplexity who is only using a pytorch model should not have to deal with jax.Arrays.   I know that the implementation for generating the data is using jax under the hood and then converting it to pytorch arrays via dlpack, but once it leaves simplexity, it should just be pure pytorch.

If you're doing this for compatibility with the logger, the better way would be to endow the logger with a last-minute converter from a tensor to numpy array I think.

File: simplexity/evaluation/evaluate_pytorch_model.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+except ImportError as e:
+    raise ImportError("To use PyTorch support install the torch extra:\nuv sync --extra pytorch") from e
+
+
+def evaluation_step(
+    model: torch.nn.Module,
+    inputs: torch.Tensor,
+    labels: torch.Tensor,
+    metric_keys: Iterable[str] = ("loss", "accuracy"),
+) -> dict[str, jax.Array]:
```
Comment (hrbigelow) [inline simplexity/evaluation/evaluate_pytorch_model.py:25]: should return dict[str, torch.Tensor]

File: simplexity/generative_processes/torch_generator.py
Comment (hrbigelow) [inline simplexity/generative_processes/torch_generator.py:1]: Here as well, why is it necessary to mix jax.Array and torch.Tensor at such a high level in the API?

File: simplexity/persistence/local_pytorch_persister.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    """Persists a PyTorch model to the local filesystem."""
+
+    filename: str = "model.pt"
+
+    def __init__(self, directory: str | Path, filename: str = "model.pt"):
+        self.directory = Path(directory)
+        self.filename = filename
+
+    # TODO: This is a hack to get the type checker to work.
+    def save_weights(self, model: torch.nn.Module, step: int = 0, overwrite_existing: bool = False) -> None:  # type: ignore
```
Comment (hrbigelow) [inline simplexity/persistence/local_pytorch_persister.py:21]: I've forgotten the semantics of torch save/load, but wasn't it considered best practice to transfer tensors to CPU before saving a checkpoint?  If so, then you wouldn't need to fuss with "map_location" when loading weights.

File: simplexity/training/train_pytorch_model.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    validation_bos_token: int | None = None,
+    validation_eos_token: int | None = None,
+) -> tuple[torch.nn.Module, float]:
+    """Train a PyTorch model on a generative process."""
+    key = jax.random.PRNGKey(training_cfg.seed)
+
+    optimizer = typed_instantiate(training_cfg.optimizer.instance, torch.optim.Optimizer, params=model.parameters())
+
+    gen_state = training_data_generator.initial_state
+    gen_states = jnp.repeat(gen_state[None, :], training_cfg.batch_size, axis=0)
```
Comment (hrbigelow) [inline simplexity/training/train_pytorch_model.py:40]: same issue here with the mixing of jax / pytorch.

File: simplexity/training/train_pytorch_model.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+        logits = model(inputs)
+
+        vocab_size = logits.shape[2]
+        logits_reshaped = logits.view(-1, vocab_size)
+        labels_reshaped = labels.view(-1).long()  # Ensure labels are long type for cross entropy
+
+        loss_tensor = F.cross_entropy(logits_reshaped, labels_reshaped)
+        loss = torch_to_jax(loss_tensor).item()
+        loss_value = loss
```
[the remark this answers is not in the document]
Comment (ealt) [reply simplexity/training/train_pytorch_model.py:66]: I only need the last loss value

File: simplexity/evaluation/evaluate_pytorch_model.py
Comment (ealt) [reply simplexity/evaluation/evaluate_pytorch_model.py:1]: At this point I only need jax for the generator, and adding something like a wrapper to the generator to accept torch tensors adds unnecessary complexity. I think using jax for the generator states is a reasonable compromise

File: tests/persistence/test_local_pytorch_persister.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+def test_local_persister(tmp_path: Path):
+    directory = tmp_path
+    filename = "test_model.pt"
+    persister = LocalPytorchPersister(directory, filename)
+    assert persister.directory == directory
+    assert persister.filename == filename
+
+    model = get_model(0)
+    assert not (tmp_path / "0" / filename).exists()
```
[the remark this answers is not in the document]
Comment (ealt) [reply tests/persistence/test_local_pytorch_persister.py:52]: That has not happened in practice
My understanding is that the tmp_path fixture is unique to each run and cleaned up after execution
